subsys/net/lib/ocpp/ocpp.c

```diff
@@ -0,0 +1,748 @@
+	FILL_PDU_TABLE(PDU_HEARTBEAT, "Heartbeat"),
+	FILL_PDU_TABLE(PDU_METER_VALUES, "MeterValues"),
+	FILL_PDU_TABLE(PDU_CLEAR_CACHE, "ClearCache"),
+	FILL_PDU_TABLE(PDU_REMOTE_START_TRANSACTION, "RemoteStartTransaction"),
+	FILL_PDU_TABLE(PDU_REMOTE_STOP_TRANSACTION, "RemoteStopTransaction"),
+	FILL_PDU_TABLE(PDU_GET_CONFIGURATION, "GetConfiguration"),
+	FILL_PDU_TABLE(PDU_CHANGE_CONFIGURATION, "ChangeConfiguration"),
+	FILL_PDU_TABLE(PDU_CHANGE_AVAILABILITY, "ChangeAvailability"),
+	FILL_PDU_TABLE(PDU_UNLOCK_CONNECTOR, "UnlockConnector"),
+	FILL_PDU_TABLE(PDU_RESET, "Reset"),
+};
+
+const char *ocpp_get_pdu_literal(enum ocpp_pdu_msg pdu)
+{
+	if (pdu >= PDU_MSG_END) {
+		return "";
+	}
+
+	return pdu_msg_table[pdu].spdu;
+}
+
+int ocpp_find_pdu_from_literal(const char *msg)
+{
+	struct ocpp_msg_table *mt = pdu_msg_table;
+	int i;
+
+	for (i = 0; i < PDU_MSG_END; i++) {
+		if (!strncmp(mt[i].spdu, msg, strlen(mt[i].spdu))) {
+			break;
+		}
+	}
+
+	return i;
+}
+
+void ocpp_get_utc_now(char utc[CISTR25])
+{
+	struct timespec tv;
+	struct tm htime = {0};
+
+	gettimeofday(&tv, NULL);
+	gmtime_r(&tv.tv_sec, &htime);
+
+	snprintk(utc, CISTR25, "%04d-%02d-%02dT%02d:%02d:%02dZ",
+		htime.tm_year + 1900,
+		htime.tm_mon + 1,
+		htime.tm_mday,
+		htime.tm_hour,
+		htime.tm_min,
+		htime.tm_sec);
+}
+
+bool ocpp_session_is_valid(struct ocpp_session *sh)
+{
+	sys_snode_t *node;
+	struct ocpp_session *lsh;
+	bool is_found = false;
+
+	if (!sh) {
+		return is_found;
+	}
+
+	k_mutex_lock(&gctx->ilock, K_FOREVER);
+	SYS_SLIST_FOR_EACH_NODE(&gctx->slist, node) {
+		lsh = to_session(node);
+		if (lsh == sh) {
+			is_found = true;
+			break;
+		}
+	}
+	k_mutex_unlock(&gctx->ilock);
+
+	return is_found;
+}
+
+static int ocpp_ws_connect_cb(int ws_sock, struct http_request *req,
+			      void *user_data)
+{
+	struct ocpp_info *ctx = user_data;
+
+	ctx->is_cs_offline = false;
+	return 0;
+}
+
+static int ocpp_connect_to_cs(struct ocpp_info *ctx)
+{
+	int ret;
+	struct websocket_request config = {0};
+	struct ocpp_upstream_info *ui = ctx->ui;
+	struct sockaddr addr4;
+	struct sockaddr_in6 addr6;
+	struct sockaddr *addr;
+	int addr_size;
+
+	if (ui->csi.sa_family == AF_INET) {
+		addr = &addr4;
+		addr_size = sizeof(addr4);
+		addr->sa_family = ui->csi.sa_family;
+		net_sin(addr)->sin_port = htons(ui->csi.port);
+		zsock_inet_pton(addr->sa_family, ui->csi.cs_ip,
+				&net_sin(addr)->sin_addr);
+	} else {
+		addr = (struct sockaddr *)&addr6;
+		addr_size = sizeof(addr6);
+		addr->sa_family = ui->csi.sa_family;
+		net_sin6(addr)->sin6_port = htons(ui->csi.port);
+		zsock_inet_pton(addr->sa_family, ui->csi.cs_ip,
+				&net_sin6(addr)->sin6_addr);
+	}
+
+	ret = zsock_connect(ui->tcpsock, addr, addr_size);
+	if (ret < 0 && errno != EALREADY && errno != EISCONN) {
+		LOG_ERR("tcp socket connect fail %d %d", ret, errno);
+		return ret;
+	}
+
+	if (ui->wssock >= 0) {
+		websocket_disconnect(ui->wssock);
+	}
+
+	if (ui->wssock < 0) {
+		char buf[128];
```

Comment:
This is a big no-no, the temp buffer is used by the websocket library for storing temporary data, if you use a variable from stack, then future websocket operations will overwrite whatever data there is in that memory.

Reply:
As far I have checked, the API call is synchronous and host buf is copied in the below websocket (http) layer. Temp buffer are dynamic allocated, please correct me if I missed something.

Reply:
Hmm, ok, I seem to have missed the fact that the buf was only used for hostname, which does not seem to cause any issues atm. It is just that the host pointer will point to invalid memory after exiting this block but fortunately the host is only used synchronously as you said. Never mind then.